### PR TITLE
[DF][ROOT-10573] Write TTree header before MT Snapshot starts 

### PR DESCRIPTION
**Pull request by eguiraud**
When the output TFile of a Snapshot is opened in UPDATE mode, and
the output TTree has the same name as a TTree already in the TFile,
the expected behavior is that Snapshot writes adds a new namecycle
to the file.

However, Snapshot+IMT is based on TBufferMerger, and TBufferMerger's
behavior in this scenario is to try to merge the new data with the
data already in the TFile (in the same namecycle).

In order to make Snapshot+IMT behave the same as a non-IMT Snapshot,
we now always write an empty TTree (with the correct branches) to
the TFile before we start an IMT fill. This first write creates the
new namecycle (initially with no data) with which TBufferMerger will
merge its buffers.

This fixes ROOT-10573.

### CODE REVIEW

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx [hunk 1]
```diff
@@ -1387,7 +1417,27 @@ public:
 
    void Initialize()
    {
+      ::TDirectory::TContext c; // do not change gDirectory outside of this scope
+
       const auto cs = ROOT::CompressionSettings(fOptions.fCompressionAlgorithm, fOptions.fCompressionLevel);
+
+      // Write an empty TTree to file as a clean starting point for TBufferMerger.
+      // This is mostly useful when a TTree with the same name is already present in the output
+      // file and we are UPDATE'ing the TFile, in which case TBufferMerger would try to merge new
```
Comment (pcanal) [inline tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx]: Should we execute this code only in the update mode case?

Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx]: It doesn't hurt to always execute it. This way the logic is always the same (easier to read, understand, follow) and I don't have to do an awkward "check if the file is there, if yes check if the tree is there".

Comment (pcanal) [reply tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx]: I was proposing "if fOptions.Mode is update then", this save on the spurious opening of file . 

Actually isn't that function also executed if the file is read-only? (and Snapshot is not called)

Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx]: There is another very good reason to always write the header: so if no event passes the selections, the file still contains an empty TTree with the proper structure. Before this change, the file would contain a TTree key with no branch information.

> Actually isn't that function also executed if the file is read-only? (and Snapshot is not called)

No, `SnapshotHelperMT::Initialize` is called right before an event loop in which a `Snapshot` has been booked.

Comment (pcanal) [reply tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx]: > No, SnapshotHelperMT::Initialize i

That this Initialize function is part of a class name SnapshotHelperMT was hard to see from github as they are many classes in the same file ... and the github expand-this-code-snippet button can expand below the snippet not above it :( .....

> There is another very good reason to always write the header: 

humm .. if the fOptions.Mode is NEW then the 2nd use later on will be wrong (lead to a Zombie file).  if the option is RECREATE then the work done here will be throw out.   So if you can going this way, then the 2nd opening of the output file must be a "UPDATE".

Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx]: > if you can going this way, then the 2nd opening of the output file must be a "UPDATE".

yes the current logic is definitely not ok :sweat: work in progress

**Comment by eguiraud** [comment]
It turns out it's quite a bit more complicated to use TTree::SetBranchAddress to reset TTree branch pointers at the beginning of every task rather than just create new branches with TTree::Branch.

In particular, TTree::Branch accepts addresses of STL collections while TTree::SetBranchAddress requires the address of a pointer to the STL collection, and the pointer is required to have same lifetime as the pointee. This would require quite a bit of book-keeping inside a MT Snapshot.

I'll open a new PR that will take a similar approach but will only rely on TTree::Branch.

**Pull request closed without merging**